doc/man/peakidentifier.1
.TH peakidentifier 1 "March 2012" "ESRF" "PyMca X-Ray Fluorescence Toolkit"

.SH NAME

peakidentifier - Given a photon energy, list the possible elements.

.SH SYNOPSIS

peakidentifier

.SH DESCRIPTION

.P
Given an energy and a threshold, list all the elements emitting X-rays 
in the range [energy-threshold, energy+threshold]. User can select the 
lines to be considered: K, L1, L2, L3, M, ... 

The program lists the element, the IUPAC line name and the relative
intensity of the line among the family of lines to which it belongs.

.SH SEE ALSO
xraylib


changelog.txt
VERSION 5.3.0

- XRF. Correct several Single Layer Strategy issues.

- XRF. Support calibration readout using HDF5 files.

- XRF. Support time readout and use of time information from HDF5 files.

- HDF5. Simplify HDF5 data selection handling by generatic automatic data selection tables
besides the user defined one (requires use of a measurement group)

- HDF5. Support readout of motor positions together with curve data when following ESRF and Sardana standards.

- HDF5. Support NXdata and default plots (requires silx)

- Add plugin to fit all curves present in a 1D window (requires h5py)

- ROI Imaging. No limit on the number of slaves.

- Add tomographic reconstruction capabilities (requires freeART and tomoGUI)

- Allow to build PyMca using GLIBC 2.26

- Activate Continuous Integration.

VERSION 5.2.2

ROI Imaging. Correct problem normalizing integer data by integer monitor.

HDF5 Widget. Use PyMca plugins on silx data viewer.

HDF5 Batch. Correct error when all the entries in an HDF5 file do not present
the same structure.

SPEC shared memory. Correct counter labelling order in case of using 10 or more
counters and motors in a scan.

NeXus. Interpret new style NXdata groups using silx NXdataViewer if silx is installed.

VERSION 5.2.1

Correct annoying error message appearing when background is not defined.

Adopt MIT license for SpecFile library

VERSION 5.2.0

RGB Correlator can now export the data as a single TIFF file or as multiple TIFF
files.

XRF: Deal with the case the mass fraction is zero in one of the compounds of a material. It can arrive when applying a Strategy.

Extend the command line usage of the FastXRFLinearFit module to HDF5 datasets.

ROI Imaging. Improved external images plugin. It requires silx.

ROI Imaging. Plugin to display per pixel information (motors and others). It
requires silx.

ROI Imaging. Allow to sum the master and the slave stack. It can be used to
sum an arbitrary number of stacks.

ROI Imaging. Selecting multiple HDF5 datasets as signal generates a stack that
is the sum of all of the datasets.

Implement a calculation cache to speed up secondary excitation calculations.

Allow to use fisx library for the calculation of escape peaks. Default is
disabled.

Improved support of OMDAQ lmf format.

Correct bug affecting the calculation of ROIs when the x axis coordinates are
negative

MacOS. Correct bug on startup when reading the default configuration.

MacOS. Frozen binary uses now HDF5 1.10.1 to support SWMR files.

VERSION 5.1.4

ROI Imaging tool can add multiple detector signals present in an HDF5 file.

Support multiple JCAMP-DX blocks in a file.

Support stacks exported by OPUS in JCAMP-DX format.

Add basic support of OMDAQ lmf data format.

Make sure main window menu options are available under MacOS and Qt5.

Support matplotlib 2.0

Correct factor of two error when fitting step up or step down functions.

VERSION 5.1.3

Correct minor Qt5 related bugs.

Simplify embedding batch fitting in custom workflows.

ROI Imaging: Exporting the stack in TIFF format did not respect zoomed region
limits

Allow exporting the x-ray tube emission continuum from the plot.

Implement a "copy-selection-to-clipboard" of the concentrations table via
CTRL-C.

Fix error message when moving mouse on SCAN window after having used the regular mesh plugin

Correct coordinate display on image window.

Solve issue calculating shell photoelectric cross-sections close to edges.

VERSION 5.1.2

Correct the generation of the efficiency plot in the ATTENUATORS tab when
BeamFilter0 was selected. That problem only affected the plot, not anything
else.

Adapt to latest SPEC version

Correct compatibility issues with matplotlib 1.5

Improved compatibility with Qt 5

Update to fisx 1.0.9 for windows compatibility under Python 3.5

Enable OpenGL window under Python 3

Implement Kaiser window option for XAS Fourier transform.

Reimplement curve renaming.

Recover interactive console functionality.

Correct the density of H and He.

Add more keywords for QXAS format support.

Allow ADD functionality on customized fit background (allowing multiple atan
functions)

Allow the use of the regular mesh plugin on incomplete mesh scans.

Add kinetics related plugins (Rate Laws and Arrhenius-like plots).

VERSION 5.1.1

Update to fisx 1.0.4 to workaround issue calculating secondary excitation
corrections when the incident beam is normal to the sample surface.

Restore pre-5.0.0 behavior in the scan window when displaying mouse
coordinates without the crosshair cursor.

Correct multiple curve averaging when input arrays are reversed.

Add the possibility to save the Monte Carlo generated matrix spectra in the
fit window.

Correct problem of reapplying the calibration when toggling log and linear
axes while already using a calibration.

Restore saving of images in png and jpg in the main window.

Restore data projection on PCA eigenvectors calculated using the covariance
method to the same way as when using the other methods.

VERSION 5.1.0

Automatic loading of user plugins.

Allow the use of user specified physical constants across different installed versions.

Add stack plugin to calculate multiple ROIs in one go.

Add basic JCAMP-DX reading support.

Add crosshair cursor option to the plot options menu.

Add EXAFS extraction capabilities to the Main Window and the ROI Imaging Tool.

Improved right axis autoscaling.

Panning with the keyboard arrow keys implemented in most 1D plots.

Correct bug on customized fit when using anchors.

Correct bug saving MCA spectra when using the Taurus plugin.

Correct swap of red and blue components when saving via matplotlib.

Correct regular mesh plotting.

VERSION 5.0.3

Recover Advanced fit graphics saving in logarithmic mode.

Slower but correct reading of SPE files.

Correctly handle Energy and Channel plot labels.

Calculate tertiary excitation in diagnostics tab.

Allow stack normalization by an external image.


PyMca5/PyMcaGui/physics/xrf/PeakIdentifier.py
#/*##########################################################################
#
# The PyMca X-Ray Fluorescence Toolkit
#
# Copyright (c) 2004-2015 European Synchrotron Radiation Facility
#
# This file is part of the PyMca X-ray Fluorescence Toolkit developed at
# the ESRF by the Software group.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
#############################################################################*/
__author__ = "V. Armando Sole - ESRF Data Analysis"
__contact__ = "sole@esrf.fr"
__license__ = "MIT"
__copyright__ = "European Synchrotron Radiation Facility, Grenoble, France"
import sys
from PyMca5.PyMcaGui import PyMcaQt as qt
from PyMca5.PyMcaPhysics import Elements
from PyMca5.PyMcaGui import PyMca_Icons
IconDict = PyMca_Icons.IconDict
QTVERSION = qt.qVersion()
DEBUG = 0


class PeakIdentifier(qt.QWidget):
    sigPeakIdentifierSignal = qt.pyqtSignal(object)
    def __init__(self,parent=None,energy=None,threshold=None,useviewer=None,
                 name="Peak Identifier"):
        qt.QWidget.__init__(self, parent)
        self.setWindowTitle(name)
        self.setWindowIcon(qt.QIcon(qt.QPixmap(IconDict['gioconda16'])))

        if energy    is None: energy    = 5.9
        if threshold is None: threshold = 0.030
        if useviewer is None: useviewer = 0
        self.__useviewer = useviewer

        layout = qt.QVBoxLayout(self)
        #heading
        self.__energyHBox=qt.QWidget(self)
        hbox = self.__energyHBox
        hbox.layout = qt.QHBoxLayout(hbox)
        hbox.layout.setContentsMargins(0, 0, 0, 0)
        hbox.layout.setSpacing(0)
        layout.addWidget(hbox)
        hbox.layout.addWidget(qt.HorizontalSpacer(hbox))

        l1=qt.QLabel(hbox)
        l1.setText('<b><nobr>Energy (keV)</nobr></b>')
        hbox.layout.addWidget(l1)
        self.energy=MyQLineEdit(hbox)
        self.energy.setText("%.3f" % energy)
        self.energy._validator = qt.QDoubleValidator(self.energy)
        self.energy.setValidator(self.energy._validator)
        self.energy.setToolTip('Press enter to validate your energy')
        hbox.layout.addWidget(self.energy)
        hbox.layout.addWidget(qt.HorizontalSpacer(hbox))
        self.energy.editingFinished.connect(self._energySlot)
        #parameters
        self.__hbox2 = qt.QWidget(self)
        hbox2 = self.__hbox2

        layout.addWidget(hbox2)
        hbox2.layout = qt.QHBoxLayout(hbox2)
        hbox2.layout.setContentsMargins(0, 0, 0, 0)
        hbox2.layout.setSpacing(0)
        font=hbox2.font()
        font.setBold(1)
        hbox2.setFont(font)


        l2=qt.QLabel(hbox2)
        l2.setText('Energy Threshold (eV)')
        self.threshold=qt.QSpinBox(hbox2)
        self.threshold.setMinimum(0)
        self.threshold.setMaximum(1000)
        self.threshold.setValue(int(threshold*1000))
        self.k = qt.QCheckBox(hbox2)
        self.k.setText('K')
        self.k.setChecked(1)
        self.l1 = qt.QCheckBox(hbox2)
        self.l1.setText('L1')
        self.l1.setChecked(1)
        self.l2 = qt.QCheckBox(hbox2)
        self.l2.setText('L2')
        self.l2.setChecked(1)
        self.l3 = qt.QCheckBox(hbox2)
        self.l3.setText('L3')
        self.l3.setChecked(1)
        self.m = qt.QCheckBox(hbox2)
        self.m.setText('M')
        self.m.setChecked(1)
        self.threshold.valueChanged[int].connect(self._thresholdSlot)
        self.k.clicked.connect(self.mySlot)
        self.l1.clicked.connect(self.mySlot)
        self.l2.clicked.connect(self.mySlot)
        self.l3.clicked.connect(self.mySlot)
        self.m.clicked.connect(self.mySlot)

        hbox2.layout.addWidget(l2)
        hbox2.layout.addWidget(self.threshold)
        hbox2.layout.addWidget(self.k)
        hbox2.layout.addWidget(self.l1)
        hbox2.layout.addWidget(self.l2)
        hbox2.layout.addWidget(self.l3)
        hbox2.layout.addWidget(self.m)

        if self.__useviewer:
            self.__browsertext = qt.QTextEdit(self)
            layout.addWidget(self.__browsertext)
        self.setEnergy(energy)

    def setEnergy(self, energy = None):
        if energy is None:
            energy = 5.9
        if type(energy) == type(""):
            self.energy.setText("%s" % energy)
            self._energySlot()
        else:
            self.energy.setText("%.3f" % energy)
            self.mySlot(energy=energy)

    def _energySlot(self):
        qstring = self.energy.text()
        try:
            value = float(qt.safe_str(qstring))
            self.energyvalue = value
            self.mySlot()
            self.energy.setPaletteBackgroundColor(qt.Qt.white)
            cursor = self.__browsertext.textCursor()
            cursor.movePosition(qt.QTextCursor.Start)
            self.__browsertext.setTextCursor(cursor)
            self.threshold.setFocus()
        except:
            msg=qt.QMessageBox(self.energy)
            msg.setIcon(qt.QMessageBox.Critical)
            msg.setText("Invalid Float")
            msg.setWindowTitle("Invalid entry")
            msg.exec_()
            self.energy.setFocus()
            return

    def myslot(self):
        print("PeakIdentifier.py myslot deprecated, use mySlot")
        return self.mySlot()

    def _thresholdSlot(self, value):
        self.mySlot()

    def mySlot(self, energy=None):
        if energy is None:
            try:
                energy  = float(qt.safe_str(self.energy.text()))
            except ValueError:
                msg=qt.QMessageBox(self.energy)
                msg.setIcon(qt.QMessageBox.Critical)
                msg.setText("Invalid Energy Value")
                msg.setWindowTitle("Invalid energy")
                msg.exec_()
                self.energy.setFocus()
                return

        threshold = float(self.threshold.value())/1000.
        lines=[]
        if self.k.isChecked():
             lines.append('K')
        if self.l1.isChecked():
             lines.append('L1')
        if self.l2.isChecked():
             lines.append('L2')
        if self.l3.isChecked():
             lines.append('L3')
        if self.m.isChecked():
             lines.append('M')
        ddict=Elements.getcandidates(energy,threshold,lines)[0]
        ddict['text'] =self.getHtmlText(ddict)
        ddict['event']='Candidates'
        ddict['lines']=lines
        if self.__useviewer:
            self.__browsertext.clear()
            #self.__browsertext.insertHtml("<CENTER>"+dict['text']+\
            #                              "</CENTER>")
            self.__browsertext.insertHtml(ddict['text'])
        self.sigPeakIdentifierSignal.emit(ddict)

    def getHtmlText(self, ddict):
        text  = ""
        if QTVERSION < '4.0.0':
            text += "<br>"
        labels=['Element','Line','Energy','Rate']
        lemmon=("#%x%x%x" % (255,250,205))
        lemmon = lemmon.upper()
        hcolor = ("#%x%x%x" % (230,240,249))
        hcolor = hcolor.upper()
        text+="<CENTER>"
        text+=("<nobr>")
        text+=( "<table WIDTH=80%%>")
        text+=( "<tr>")
        for l in labels:
            text+=('<td align="left" bgcolor="%s"><b>' % hcolor)
            text+=l
            text+=("</b></td>")
        text+=("</tr>")
        for ele in ddict['elements']:
            oldline=""
            for line in ddict[ele]:
                if   line[0][0:1] == 'K':
                    group0 = 'K  rays'
                elif line[0][0:2] == 'L1':
                    group0 = 'L1 rays'
                elif line[0][0:2] == 'L2':
                    group0 = 'L2 rays'
                elif line[0][0:2] == 'L3':
                    group0 = 'L3 rays'
                elif line[0][0:1] == 'M':
                    group0 = 'M rays'
                else:
                    group0 = 'Unknown'
                if group0 != oldline:
                    text +="<tr>"
                    text += '<td align="left"><b>%s</b></td>' % ele
                    text += '<td align="left"><b>%s</b></td>' % group0
                    text += '</tr>'
                    oldline = group0
            #for peak in result[group]['peaks']:
                text += '<tr><td></td>'
                name   = line[0]
                energy = ("%.3f" % line[1])
                ratio  = ("%.5f" % line[2])
                fields = [name,energy,ratio]
                for field in fields:
                    if field == name:
                        text+=('<td align="left"  bgcolor="%s">%s</td>' % (lemmon,field))
                    else:
                        text+=('<td align="right" bgcolor="%s">%s</td>' % (lemmon,field))
                text+="</tr>"
        text+=("</table>")
        text+=("</nobr>")
        text+="</CENTER>"
        return text

class MyQLineEdit(qt.QLineEdit):
    def __init__(self,parent=None,name=None):
        qt.QLineEdit.__init__(self,parent)
        self.setAutoFillBackground(True)

    def setPaletteBackgroundColor(self, color):
        palette = qt.QPalette()
        role = self.backgroundRole()
        palette.setColor(role,color)
        self.setPalette(palette)


    def focusInEvent(self,event):
        self.setPaletteBackgroundColor(qt.QColor('yellow'))
        # TODO not like focusOutEvent ?
        '''
        if QTVERSION > '4.0.0':
            qt.QLineEdit.focusInEvent(self, event)
        '''

    def focusOutEvent(self,event):
        self.setPaletteBackgroundColor(qt.QColor('white'))
        qt.QLineEdit.focusOutEvent(self, event)

def main():
    app  = qt.QApplication(sys.argv)
    winpalette = qt.QPalette(qt.QColor(230,240,249),qt.QColor(238,234,238))
    app.setPalette(winpalette)
    if len(sys.argv) > 1:
        ene = float(sys.argv[1])
    else:
        ene = 5.9
    mw = qt.QWidget()
    l  = qt.QVBoxLayout(mw)
    l.setSpacing(0)
    w= PeakIdentifier(mw,energy=ene,useviewer=1)
    l.addWidget(w)
    mw.setWindowTitle("Peak Identifier")
    mw.show()
    app.exec_()

if __name__ == "__main__":
    main()
